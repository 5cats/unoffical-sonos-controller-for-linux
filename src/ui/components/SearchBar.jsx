import _ from 'lodash';
import React, { Component } from 'react';
import { connect } from 'react-redux';
import shallowCompare from 'shallow-compare';

import SearchBarSources from './SearchBarSources';

import { search, exitSearch } from '../reduxActions/BrowserListActions';

import {
    getCurrentState,
    getSearching,
    getSearchMode,
    getSearchSources
} from '../selectors/BrowserListSelectors';

function mapStateToProps(state) {
    return {
        term: state.browserList.searchTerm,
        currentState: getCurrentState(state),
        searching: getSearching(state),
        searchMode: getSearchMode(state),
        sources: getSearchSources(state)
    };
}

function mapDispatchToProps(dispatch) {
    return {
        search: (term, mode) => dispatch(search(term, mode)),
        exitSearch: () => dispatch(exitSearch())
    };
}

export class SearchBar extends Component {
    constructor() {
        super();

        this.ref = React.createRef();

        this.inputHandler = _.debounce(this._onChange.bind(this), 800, {
            trailing: true,
            leading: false
        });
    }

    shouldComponentUpdate(nextProps, nextState) {
        return shallowCompare(this, nextProps, nextState);
    }

    UNSAFE_componentWillReceiveProps(props) {
        if (!props.term || props.term === '') {
            this.ref.current.value = '';
        }
    }

    render() {
        let cancelButton;

        if (this.props.searching) {
            cancelButton = (
                <i
                    className="material-icons"
                    onClick={this._onClick.bind(this)}
                >
                    cancel
                </i>
            );
        }

        return (
            <div id="search">
                <SearchBarSources {...this.props} />
                <input
                    ref={this.ref}
                    type="text"
                    id="searchfield"
                    onChange={this.inputHandler}
                />
                {cancelButton}
            </div>
        );
    }

    _onClick() {
        this.props.exitSearch();
        this.ref.current.value = '';
    }

    _onChange(e) {
        const term = this.ref.current.value;
        this.props.search(term, this.props.searchMode);
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(SearchBar);
