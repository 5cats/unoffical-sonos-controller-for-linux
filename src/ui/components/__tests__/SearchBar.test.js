import React from 'react';
import { SearchBar } from '../SearchBar';
import { deep } from 'react-render-spy';

describe('SearchBar', () => {
    it('renders on', () => {
        const props = {
            sources: [],
            currentState: {}
        };
        const context = deep(<SearchBar {...props} />);
        expect(context.output()).toMatchSnapshot();
    });
});
