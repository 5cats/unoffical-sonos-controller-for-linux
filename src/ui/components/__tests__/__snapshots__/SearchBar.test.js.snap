// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar renders on 1`] = `
<div
  id="search"
>
  <p />
  <input
    id="searchfield"
    type="text"
  />
</div>
`;
